File: W9_Synchronization/monitor/ProducerConsumer.java
/**
 * This is the sample program in page 138,
 *  Andrew S. Tanenbaum and Herbert Bos. 2014. Modern Operating Systems (4th
 *  ed.). Prentice Hall Press, Upper Saddle River, NJ, USA. 
 */
import java.util.Random;

public class ProducerConsumer {
    static final int N = 100;
    static Producer p = new Producer();
    static Consumer c = new Consumer();
    static ProducerConsumerMonitor monitor = new ProducerConsumerMonitor();

    public static void main(String[] args) {
        p.start();
        c.start();
    }

    static class Producer extends Thread {

        @Override
        public void run() {
            int item;
            while (true) {
                item = produceItem();
                monitor.insert(item);
            }
        }

        private int produceItem() {
            Random rng = new Random();
            return rng.nextInt();
        }
    }


    static class Consumer extends Thread {

        @Override
        public void run() {
            int item;
            while (true) {
                item = monitor.remove();
                consumeItem(item);
            }
        }

        private void consumeItem(int item) {
        }
    }

    static class ProducerConsumerMonitor {
        private int buffer[] = new int[N];
        private int count = 0, low = 0, high = 0;

        public synchronized void insert(int val) {
            if (count == N) {      // buffer is full, sleep while waiting
                goToSleep();
            }
            buffer[high] = val;    // insert the produced item to buffer
            high = (high + 1) / N; // advance to next available slot
            count = count + 1;     // count the number of items in buffer
            if (count == 1) {      
                // buffer is not empty any more, wake up consumer to consume
                notify();
            }
        }

        public synchronized int remove() {
            int val;

            if (count == 0) {
                goToSleep();
            }
            val = buffer[low];
            low = (low + 1) % N;
            count = count - 1;
            if (count == N - 1) { 
                // buffer is not full any more, wake up producer to produce
                notify();
            }
            return val;
        }

        private void goToSleep() {
            try {
                wait(); 
            } catch (InterruptedException e) {
            }
        }
    }
}

File: C19aSynchronization/mutex/producerconsumer.c
/**
 * This is the sample program in page 138,
 *  Andrew S. Tanenbaum and Herbert Bos. 2014. Modern Operating Systems (4th
 *  ed.). Prentice Hall Press, Upper Saddle River, NJ, USA. 
 */

#include <stdio.h> 
#include <stdlib.h> 
#include <pthread.h>

#define MAX 1000
pthread_mutex_t the_mutex;
pthread_cond_t condc, condp;
int buffer = 0;

void *producer(void *ptr) {
    int i;

    for (i=1; i<=MAX; i++) {
        pthread_mutex_lock(&the_mutex);
        while (buffer != 0) {
            pthread_cond_wait(&condp, &the_mutex);
        }
        buffer = i;
        pthread_cond_signal(&condc);
        pthread_mutex_unlock(&the_mutex);
    }
    pthread_exit(0);
}

void *consumer(void *ptr) {
    int i;

    for (i=1; i<=MAX; i++) {
        pthread_mutex_lock(&the_mutex);
        while (buffer == 0) {
            pthread_cond_wait(&condc, &the_mutex);
        }
        buffer = 0;
        pthread_cond_signal(&condp);
        pthread_mutex_unlock(&the_mutex);
    }
    pthread_exit(0);
}

int main(int argc, char *argv[]) {
    pthread_t producer_thread, consumer_thread;

    pthread_mutex_init(&the_mutex, 0);
    pthread_cond_init(&condc, 0);
    pthread_cond_init(&condp, 0);

    pthread_create(&consumer_thread, 0, consumer, 0);
    pthread_create(&producer_thread, 0, producer, 0);
    
    pthread_join(producer_thread, 0);
    pthread_join(consumer_thread, 0);

    pthread_cond_destroy(&condc);
    pthread_cond_destroy(&condp);
    pthread_mutex_destroy(&the_mutex);

    return EXIT_SUCCESS;
}

